src/celeritas/track/detail/TrackSortUtils.cu

```diff
@@ -65,6 +67,89 @@ void sort_impl(TrackSlots const& track_slots, F&& func)
     CELER_DEVICE_CHECK_ERROR();
 }
 
+template<class F>
+CELER_FUNCTION void
+tracks_per_action_impl(DeviceRef<CoreStateData> const& states,
+                       F&& action_accessor)
+{
+    ThreadId tid = celeritas::KernelParamCalculator::thread_id();
+
+    if ((tid < states.size()) && tid.get() != 0)
+    {
+        Span<ThreadId> offsets = states.thread_offsets[AllItems<ThreadId>{}];
+        ActionId current_action = action_accessor(tid);
+        ActionId previous_action = action_accessor(tid - 1);
+        if (current_action && current_action != previous_action)
+        {
+            offsets[current_action.get()] = tid;
+        }
+    }
+}
+
+// PRE: action_accessor is sorted, i.e. i <= j ==> action_accessor(i) <=
+// action_accessor(j)
+__global__ void tracks_per_action_kernel(DeviceRef<CoreStateData> const states,
+                                         TrackOrder order)
+{
+    // DISPATCH here since CELER_LAUNCH_KERNEL doesn't work with templated
+    // kernels
+    switch (order)
+    {
+        case TrackOrder::sort_along_step_action:
+            return tracks_per_action_impl(
+                states,
+                along_step_action_accessor{states.sim.along_step_action.data(),
+                                           states.track_slots.data()});
+        case TrackOrder::sort_step_limit_action:
+            return tracks_per_action_impl(
+                states,
+                step_limit_action_accessor{states.sim.step_limit.data(),
+                                           states.track_slots.data()});
+        default:
+            CELER_ASSERT_UNREACHABLE();
+    }
+}
+
+// TODO: On host
+__global__ void
+tracks_per_action_reduce_kernel(DeviceRef<CoreStateData> const states)
+{
+    ThreadId tid = celeritas::KernelParamCalculator::thread_id();
+    if (tid.get() == 0)
+    {
+        Span<ThreadId> offsets = states.thread_offsets[AllItems<ThreadId>{}];
+        offsets.back() = ThreadId{states.size()};
+        for (auto thread_id = offsets.end() - 2; thread_id >= offsets.begin();
+             --thread_id)
+        {
+            if (*thread_id == ThreadId{})
+            {
+                *thread_id = *(thread_id + 1);
+            }
+        }
+    }
+}
+
+void tracks_per_action(DeviceRef<CoreStateData> const& states, TrackOrder order)
+{
+    auto start = device_pointer_cast(states.thread_offsets.data());
+    thrust::fill(start, start + states.thread_offsets.size(), ThreadId{});
+    CELER_DEVICE_CHECK_ERROR();
+
+    CELER_LAUNCH_KERNEL(tracks_per_action,
+                        celeritas::device().default_block_size(),
+                        states.size(),
+                        states,
+                        order);
+    CELER_DEVICE_CHECK_ERROR();
+
+    CELER_LAUNCH_KERNEL(tracks_per_action_reduce,
+                        celeritas::device().default_block_size(),
+                        states.size(),
+                        states);
+    CELER_DEVICE_CHECK_ERROR();
```

Comment:
@sethrj I think we have to split that in multiple kernel calls since each step (fill, find offset, backfill) can access data written by a thread in an arbitrary block. Last kernel in not very efficient... the other option would be to transfer data back and forth between host and device which is not much better..